src/Lean/Meta/ProofBelow.lean

```diff
@@ -0,0 +1,373 @@
+      forallTelescope motive fun xs _ => do
+      mkLambdaFVars xs $ mkAppN (mkConst ctx.proofBelowNames[idx] levelParams) $ (params ++ motives ++ xs)
+    withMVarContext m do
+    let recursorInfo ← getConstInfo $ indVal.name ++ "rec"
+    let recLevels := 
+      if recursorInfo.numLevelParams > levelParams.length 
+      then levelZero::levelParams 
+      else levelParams 
+    let recursor ← mkAppN (mkConst recursorInfo.name $ recLevels) $ params ++ motives
+    apply m recursor 
+  
+  applyCtors (ms : List MVarId) : MetaM $ List MVarId := do
+    let mss ← ms.toArray.mapIdxM fun idx m => do
+      let m ← introNPRec m 
+      (←getMVarType m).withApp fun proofBelow args =>
+      args.back.withApp fun ctor args =>
+      let ctor := ctor.constName!.updatePrefix proofBelow.constName!
+      withMVarContext m do
+      let ctor := mkConst ctor proofBelow.constLevels!
+      apply m ctor 
+    return mss.foldr List.append []
+
+  introNPRec (m : MVarId) : MetaM MVarId := do
+    if (←getMVarType m).isForall then introNPRec (←intro1P m).2 else m
+  
+  closeGoal (vars : BrecOnVariables) (m : MVarId) : MetaM Unit := do
+    unless ←isExprMVarAssigned m do
+      let m ← introNPRec m
+      unless ←solveByAssumption m do
+        let subGoals ← applyIH m vars
+        subGoals.forM (closeGoal vars)
+  
+  solveByAssumption (m : MVarId) : MetaM Bool := do
+    withMVarContext m do
+    let lctx ← getLCtx 
+    let mTy ← getMVarType m
+    lctx.anyM fun localDecl => do
+      let (mvars, _, t) ← forallMetaTelescope localDecl.type
+      commitWhen do
+      if ←isDefEq mTy t then 
+        assignExprMVar m t
+        if ←mvars.allM (fun v => isExprMVarAssigned v.mvarId!) then 
+          assignExprMVar m (mkAppN (mkFVar localDecl.fvarId) mvars)
+          true
+        else false
+      else false
+
+def mkBrecOnDecl (ctx : Context) (idx : Nat) : MetaM Declaration := do
+  let type ← mkType
+  let indVal := ctx.typeInfos[idx]
+  let name := indVal.name ++ brecOnSuffix
+  Declaration.thmDecl { 
+    name := indVal.name ++ brecOnSuffix
+    levelParams := indVal.levelParams
+    type := type
+    value := ←proveBrecOn ctx indVal type }
+where
+  mkType : MetaM Expr :=
+    forallTelescope ctx.headers[idx] fun xs t => do
+    let params := xs[:ctx.numParams]
+    let motives := xs[ctx.numParams:ctx.numParams + ctx.motives.size].toArray
+    let indices := xs[ctx.numParams + ctx.motives.size:]
+    let motiveBinders ← ctx.motives.mapIdxM $ mkIH params motives
+    withLocalDeclsD motiveBinders fun ys => do
+    mkForallFVars (xs ++ ys) (mkAppN motives[idx] indices)
+  mkIH 
+      (params : Array Expr)
+      (motives : Array Expr)
+      (idx : Fin ctx.motives.size) 
+      (motive : Name × Expr) : MetaM $ Name × (Array Expr → MetaM Expr) := do
+    let name :=
+      if ctx.motives.size > 1
+      then mkFreshUserName s!"ih_{idx.val.succ}"
+      else mkFreshUserName "ih"
+    let ih ← instantiateForall motive.2 params
+    let mkDomain _ := 
```

Comment:
Why does `mkDomain` take an `Array Expr` if it doesn't use it?

Reply:
the culprit is: 
```lean
let motiveBinders ← ctx.motives.mapIdxM $ mkIH params motives
withLocalDeclsD motiveBinders fun ys => do
```

since `withLocalDeclsD` allows for dependent binders, it needs to take in an `Array Expr`. But at the same time, this is an independent binder. So I take the args but ignore them.